support other data type than string, add some failsafes

BigRoy opened the pull request:
Update from current pypeclub:master

mkolar committed:
support other data type than string, add some failsafes

BigRoy commented:
@mkolar is this the latest version you are using in production?

I noticed #2 is still an open PR and never was merged - are you using that code? If I merge that it might break your usages so maybe we'll need to iterate on what changes make sense currently for acre?

mkolar commented:
To be honest I need to double check with the team. I believe we tried to go back to your original code, but I lost track now. Will get back to you asap

antirotor committed:
fixing partial formatting

antirotor committed:
add pyproject.toml

antirotor committed:
add setup.py shim

antirotor committed:
fix toml

antirotor committed:
added wheel

antirotor committed:
add setup.cfg

mkolar committed:
Merge pull request #1 from antirotor/fix/unformatted-tokens

partial formatting

mkolar committed:
Merge pull request #2 from antirotor/feature/add-build-system

add build files

antirotor committed:
fix failure to add path

this will allow duplicate paths but will also solve an issue with not appending already related paths

iLLiCiTiT committed:
fix fix to avoid duplicated entries in enviornments

iLLiCiTiT committed:
skip url paths before append

iLLiCiTiT committed:
have ability to modify missing key for formattings

iLLiCiTiT committed:
do not use append after merge

iLLiCiTiT committed:
do not skip values that are converted to string

iLLiCiTiT committed:
append handle empty string values

iLLiCiTiT committed:
Merge pull request #3 from pypeclub/feature/merge_without_append

Merge without append

antirotor committed:
fix license field in pyproject

BigRoy commented on acre/core.py:
```diff
@@ -44,14 +44,17 @@ def compute(env,
-            if dependency == key:
-                continue
+        try:
+            dependent_keys = re.findall("{(.+?)}", value)
+            for dependency in dependent_keys:
+                # Ignore direct references to itself because
+                # we don't format with itself anyway
+                if dependency == key:
+                    continue
 
-            dependencies.append((key, dependency))
+                dependencies.append((key, dependency))
+        except Exception:
+            dependencies.append((key, value))
```
Why does this need to be a try..except? Doesn't make much sense to me.

That'd only occur if `value` is not a string maybe? E.g. a number or None? But that basically means the `env` isn't an "environment" so the input is already invalid to begin with? This **should** error. No?

BigRoy commented on acre/core.py:
```diff
@@ -66,13 +69,17 @@ def compute(env,
     # Format dynamic values
     for key in reversed(result.sorted):
         if key in env:
+            if not isinstance(env[key], str):
+                continue
```
What kind of values are you expected in the input `env`. I'd say that anything that is not a string is not an "environment value" and thus is invalid input. That should not be continued but **should** raise an error. It means the input value is wrong.

Unless maybe we'd want to explicitly skip None values?

BigRoy commented on acre/lib.py:
```diff
@@ -37,12 +38,24 @@ class FormatDict(dict):
-    return formatter.vformat(s, (), mapping)
+    try:
+        f = formatter.vformat(s, (), mapping)
+    except Exception:
+        r_token = re.compile(r"({.*?})")
+        matches = re.findall(r_token, s)
+        f = s
+        for m in matches:
+            try:
+                f = re.sub(m, m.format(**data), f)
+            except KeyError:
+                continue
+    return f
```
Could we improve the Exception (e.g. more explicit; what exception should we capture?) and add a comment to the code as to what this is trying to solve? When should this exception occur?

BigRoy reviewed (commented):
Some general notes about the code.

antirotor committed:
update license key round 2

antirotor committed:
PEP 621

antirotor committed:
PEP 621 another round

antirotor committed:
missing required version

BigRoy committed:
Ignore values that don't have correct braces to e.g. allow some linux BASH_FUNCTIONs to exist in environment

iLLiCiTiT committed:
Merge pull request #4 from BigRoy/bugfix/linux_bash_functions

Ignore values with invalid curly braces or other invalid characters in the formatting
